util/generate-rsa-signing-tests.py

```diff
@@ -40,57 +56,163 @@ def parse(fn):
+    print 'Sig = %s' % case['S']
+    print 'Result = %s' % case['Result']
+    print ''
+
 
+def main(fn, test_type, padding_alg):
+    # File header
+    print "# NIST RSA PSS Test Vectors for FIPS 186-4 from %s in\n" % fn + \
+    "# http://csrc.nist.gov/groups/STM/cavp/documents/dss/186-3rsatestvectors.zip\n" + \
+    "# accessible from" + \
+    "# http://csrc.nist.gov/groups/STM/cavp/digital-signatures.html#test-vectors.\n" + \
+    "# Generated with the help of the util/generate-rsa-signing-tests.py script.\n" + \
+    "#\n" + \
+    "# Digest = SHAAlg.\n" + \
+    "# Key is a DER-formatted PKCS1-encoded RSA key.\n" + \
+    "# Sig = S.\n\n"
+
+    num_cases = 0
+
+    # Each test type has a different field as the last entry per case
+    # For verify tests,PKCS "Result" is always the last field.
+    # Otherwise, for signing tests, it is dependent on the padding used.
+    if test_type == "verify":
+        last_field = "Result"
+    elif padding_alg == "PSS":
+        last_field = "SaltVal"
+    else:
+        last_field = "S"
+
+    for case in parse(fn, last_field):
         if case['SHAAlg'] == 'SHA224':
             # SHA224 not supported in *ring*.
+            debug("Skipping due to use of SHA224", DEBUG)
+            continue
+
+        if case['SHAAlg'] == 'SHA1' and padding_alg == 'PSS':
+            # SHA1 with PSS not supported in *ring.
+            debug("Skipping due to use of SHA1 and PSS.", DEBUG)
             continue
 
+        # We only support PSS where the salt length is equal to the output
+        # length of the hash algorithm
+        if padding_alg == "PSS":
+            if len(case['SaltVal']) * 2 != DIGEST_OUTPUT_LENGTHS[case['SHAAlg']]:
+                debug("Skipping due to incorrect salt length.", DEBUG)
+                continue
+
         # Read private key components.
         n = int(case['n'], 16)
         e = int(case['e'], 16)
         d = int(case['d'], 16)
 
-        # Recover the prime factors and CRT numbers.
-        p, q = rsa.rsa_recover_prime_factors(n, e, d)
-        # cryptography returns p, q with p < q by default. *ring* requires
-        # p > q, so swap them here.
-        p, q = max(p, q), min(p, q)
-        dmp1 = rsa.rsa_crt_dmp1(d, p)
-        dmq1 = rsa.rsa_crt_dmq1(d, q)
-        iqmp = rsa.rsa_crt_iqmp(p, q)
-
-        # Create a private key instance.
-        pub = rsa.RSAPublicNumbers(e, n)
-        priv = rsa.RSAPrivateNumbers(p, q, d, dmp1, dmq1, iqmp, pub)
-        key = priv.private_key(default_backend())
-
-        # Recalculate and compare the signature to validate our processing.
-        msg = case['Msg'].decode('hex')
-        sig = key.sign(msg, padding.PKCS1v15(),
-                       getattr(hashes, case['SHAAlg'])())
-        hex_sig = ''.join('{:02x}'.format(ord(c)) for c in sig)
-        assert hex_sig == case['S']
+        if n.bit_length() < 2048:
+            debug("Skipping due to modulus length (too small).", DEBUG)
+            continue
```

Comment:
This check should be `n.bit_length() // 8 < 2048 // 8`. In particular, *ring* considers any modulus where the length in bytes * 8 == 2048 to be large enough.